index.js

```diff
@@ -94,7 +97,36 @@ function parseChar(character, state) {
     } else {
       state.escaped = false;
     }
-  } else if (state.regexp) {
+    return end();
+  } else if (state.backTickDepth) {
+    if (character === '$' && !state.escaped) {
+      state.dollarSign = true;
+      return end();
+    } else if (character === '{' && state.dollarSign && !state.escaped) {
+      state.interpolationDepth++;
+      state.dollarSign = false;
+      return end();
+    } else {
+      state.dollarSign = false;
+      if (character === '\\' && !state.escaped) {
+        state.escaped = true;
+        return end();
+      } else if (character === '`' && !state.escaped) {
+        if (state.interpolationDepth >= state.backTickDepth) {
+          state.backTickDepth++;
+        } else {
+          state.backTickDepth--;
+        }
+        return end();
+      } else if (character === '}' && !state.escaped && state.interpolationDepth === state.backTickDepth) {
+        state.interpolationDepth--;
+        return end();
+      } else {
+        state.escaped = false;
+      }
+    }
+  }
```

Comment:
Could some of these options get triggered accidentally inside a regexp that's nested inside a template literal?  I think maybe the best bet is to re-write some of the logic so that instead of tracking nesting via a load of separate properties, we instead should keep an array to denote the stack of nesting.  This would have the advantage/disadvantage of enforcing proper nesting i.e. it would not accept `{[}]` as being valid/well nested.